src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,187 @@
+
+export { Logger } from "../relayFeeCalculator"; // Permit adapters to import a common definition.
+export type TokenPrice = CoinGeckoPrice; // Temporary inversion; CoinGecko should source from here.
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getPriceByAddress(address: string, currency: string): Promise<TokenPrice>;
+  getPricesByAddress(addresses: string[], currency: string): Promise<TokenPrice[]>;
+}
+
+// It's convenient to map TokenPrice objects by their address, but consumers typically want an array
+// of TokenPrice objects, so the address must also be embedded within the TokenPrice object. To
+// avoid storing multiple copies of the same TokenPrice, always use the lower-case variant of the
+// address when performing a lookup into the PriceCache, and substitute the provided address into the
+// TokenPrice object when returning to the caller.
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  protected prices: {
+    [currency: string]: PriceCache;
+  } = {};
+
+  constructor(protected logger: Logger, readonly priceFeeds: PriceFeedAdapter[]) {
+    assert(logger, "No logging instance supplied.");
+    assert(priceFeeds.length > 0, "No price feeds supplied.");
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    assert(this.priceFeeds.length > 0, "No price feeds are registered.");
+    const priceCache: PriceCache = this.getPriceCache(currency);
+    const now: number = msToS(Date.now());
+
+    let tokenPrice: TokenPrice | undefined = priceCache[address.toLowerCase()];
+    const cacheMiss = tokenPrice === undefined || now - this.maxPriceAge > tokenPrice.timestamp;
+
+    if (this.maxPriceAge > 0) {
+      const age: number = tokenPrice ? now - tokenPrice.timestamp : Number.MAX_SAFE_INTEGER;
+      this.logger.debug({
+        at: "PriceClient#getPriceByAddress",
+        message: `Cache ${cacheMiss ? "miss" : "hit"} (${currency}) for token ${address}.`,
+        age: `${age} S`,
+        price: tokenPrice,
+      });
+    }
+
+    if (cacheMiss) {
+      const prices: TokenPrice[] = await this.getPricesByAddress([address], currency);
+      tokenPrice = prices[0];
```

Comment:
Just if you want to avoid a temp variable.
```suggestion
      ([tokenPrice] = await this.getPricesByAddress([address], currency));
```